### PR TITLE
[TH-379] Change update rate exchange name and add postfix to queues

### PR DESCRIPTION
**Pull request by RobMKR**
This change is [Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)

### CODE REVIEW

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 1]
```diff
@@ -47,25 +51,30 @@ public function addExchangeRateUpdateListener(ExchangeRateUpdateListener $listen
...
 				if ($this->prefetchCount === null) {
 					//Should return something, but the value is not used in next hanndler anyway...
 					return Promise\resolve();
 				}
 				return $channel->qos(0, $this->prefetchCount);
-			})->then(function() use($channel) {
+			})->then(function() use($channel, $queueName) {
 				$consumers = [];
-				$consumers[] = $channel->consume([$this, 'updateExchangeRate'], 'usd-exchange-rate.update', 'usd-exchange-rate.update');
+				$consumers[] = $channel->consume([$this, 'updateExchangeRate'], $queueName, $queueName);
```
Comment (foowie) [inline src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:77]: Can we not use consumer tag here?

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 2]
```diff
@@ -19,12 +19,16 @@
 class RabbitMqExchangeRateUpdater implements LoggerAwareInterface {
 	use LoggerAwareTrait;
 
-	const WEB_RPC_EXCHANGE = 'bitoasis.web-events';
+	const WEB_RPC_EXCHANGE = 'bitoasis.fiat-rates';
```
Comment (foowie) [inline src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:22]: Constant name is confusing. Same for other below

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 3]
```diff
@@ -19,12 +19,16 @@
...
 	const WEB_RPC_EXCHANGE_TYPE = 'direct';
+	const WEB_RPC_QUEUE_NAME = 'usd-exchange-rate.update';
 
 	/** @var int|null */
 	protected $prefetchCount = 25;
 
+	/** @var string */
+	protected $queuePostfix = '';
```
Comment (foowie) [inline src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:30]: Why is this needed?

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php (hunk 2 above)
Comment (RobMKR) [reply src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:22]: Changed

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php (hunk 1 above)
Comment (RobMKR) [reply src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:77]: Removed

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php (hunk 3 above)
Comment (RobMKR) [reply src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:30]: If the queue will be same for both orderbook and trade, only one of them will consume message from queue :) So added a postfix to create different queues for different microservices. 

As of now, orderbook will create queue `usd-exchange-rate.update.orderbook` and trades will create `usd-exchange-rate.update.trades`. One message => 2 queues => 2 consumers

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 4]
```diff
@@ -48,14 +52,13 @@ public function addExchangeRateUpdateListener(ExchangeRateUpdateListener $listen
...
 
 	public function initializeMq(Channel $channel): PromiseInterface {
-		return $channel->exchangeDeclare(self::WEB_RPC_EXCHANGE, self::WEB_RPC_EXCHANGE_TYPE, false, true)
+		return $channel->exchangeDeclare(self::FIAT_RATES_EXCHANGE, self::FIAT_RATES_EXCHANGE_TYPE, false, true)
 			->then(function() use($channel) {
-				$queues = [];
-				$queues[] = $channel->queueDeclare('usd-exchange-rate.update', false, true);
-				return Promise\all($queues);
-			})->then(function() use($channel) {
+				return $channel->queueDeclare('', false, true, true);
```
Comment (foowie) [inline src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:57]: Why keeping it durable?

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 5]
```diff
@@ -19,12 +20,15 @@
 class RabbitMqExchangeRateUpdater implements LoggerAwareInterface {
 	use LoggerAwareTrait;
 
-	const WEB_RPC_EXCHANGE = 'bitoasis.web-events';
-	const WEB_RPC_EXCHANGE_TYPE = 'direct';
+	const FIAT_RATES_EXCHANGE = 'bitoasis.fiat-rates';
+	const FIAT_RATES_EXCHANGE_TYPE = 'direct';
```
Comment (foowie) [inline src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:24]: Can you please elaborate why direct or why fanout?

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php (hunk 4 above)
Comment (RobMKR) [reply src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:57]: Didn't think about it. Agree, transient also will work fine for current task :)

File: src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php [hunk 6]
```diff
@@ -48,14 +52,13 @@ public function addExchangeRateUpdateListener(ExchangeRateUpdateListener $listen
...
-		return $channel->exchangeDeclare(self::WEB_RPC_EXCHANGE, self::WEB_RPC_EXCHANGE_TYPE, false, true)
+		return $channel->exchangeDeclare(self::FIAT_RATES_EXCHANGE, self::FIAT_RATES_EXCHANGE_TYPE, false, true)
 			->then(function() use($channel) {
-				$queues = [];
-				$queues[] = $channel->queueDeclare('usd-exchange-rate.update', false, true);
-				return Promise\all($queues);
-			})->then(function() use($channel) {
+				return $channel->queueDeclare('', false, true, true);
+			})->then(function(MethodQueueDeclareOkFrame $queueDeclareReply) use ($channel) {
+				$this->queueName = $queueDeclareReply->queue;
```
[the remark this answers is not in the document]
Comment (RobMKR) [reply src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php:59]: Correct

**Review by foowie — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: all files reviewed, 1 unresolved discussion (waiting on @RobMKR)

___
*[src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php, line 56 at r3](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12#-MxVaahP7adFwhnrMOo5:-MxVaahP7adFwhnrMOo6:b-aoj4d1) ([raw file](https://github.com/bit-oasis/bitfinex-api/blob/8fb75c9d74c23707ba682db92662f17f594c9d74/src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php#L56)):*
> ```PHP
> 	public function addRoutingKey(string $routingKey): void {
> 		if (!isset($this->routingKeys[$routingKey])) {
> 			$this->routingKeys[$routingKey] = $routingKey;
> ```

It will be good to
1) validate if RMQ is already running; if yes, throw an exception
2) I'm not sure if this should be lower case or upper case - so best to unify it here

**Review by foowie — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: all files reviewed, 1 unresolved discussion (waiting on @RobMKR)

___
*[src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php, line 56 at r3](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12#-MxVaahP7adFwhnrMOo5:-MxVgbju3Qm1C6RPQrvg:b-hsqpki) ([raw file](https://github.com/bit-oasis/bitfinex-api/blob/8fb75c9d74c23707ba682db92662f17f594c9d74/src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php#L56)):*
<details><summary><i>Previously, foowie (Daniel Robenek) wrote…</i></summary><blockquote>

It will be good to
1) validate if RMQ is already running; if yes, throw an exception
2) I'm not sure if this should be lower case or upper case - so best to unify it here
</blockquote></details>

Why to name it "addRoutinKey" in the first place? Do external classes need to know that routing key is being added or its only internal thing? Wouldn't it make more sense to name this method in a way that it would describe its behavior - that updates for given currency will be requested?

**Review by RobMKR — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @foowie)

___
*[src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php, line 56 at r3](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12#-MxVaahP7adFwhnrMOo5:-MxYqSQ157wL0vNFRqk_:b-36gsh4) ([raw file](https://github.com/bit-oasis/bitfinex-api/blob/8fb75c9d74c23707ba682db92662f17f594c9d74/src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php#L56)):*
<details><summary><i>Previously, foowie (Daniel Robenek) wrote…</i></summary><blockquote>

Why to name it "addRoutinKey" in the first place? Do external classes need to know that routing key is being added or its only internal thing? Wouldn't it make more sense to name this method in a way that it would describe its behavior - that updates for given currency will be requested?
</blockquote></details>

Maybe I didn't understand what you meant by check RMQ is running or no, because it is already running always, as we are passing RMQ channgel from another microservice. I just checked if exchange declaration is done, throw an excpetion

**Review by foowie — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @RobMKR)

**Review by foowie — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: all files reviewed, 1 unresolved discussion (waiting on @RobMKR)

___
*[src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php, line 66 at r4](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12#-MycJq4C9D1NaGQye4PU:-MycJq4DAx9G_D4P8mlt:b4y987u) ([raw file](https://github.com/bit-oasis/bitfinex-api/blob/b68f04173368e1586ed3a54cfe5590e489726138/src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php#L66)):*
> ```PHP
> 		}
> 
> 		$routingKey = strtoupper($bitOasisCurrency);
> ```

Just checked and in all our places routing keys are all lower case. Can you please just fix this last thing?

**Review by RobMKR — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @foowie)

___
*[src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php, line 66 at r4](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12#-MycJq4C9D1NaGQye4PU:-MygWJWH8o4m86oAVYrt:b499q7j) ([raw file](https://github.com/bit-oasis/bitfinex-api/blob/b68f04173368e1586ed3a54cfe5590e489726138/src/BitOasis/Bitfinex/ExchangeRate/RabbitMqExchangeRateUpdater.php#L66)):*
<details><summary><i>Previously, foowie (Daniel Robenek) wrote…</i></summary><blockquote>

Just checked and in all our places routing keys are all lower case. Can you please just fix this last thing?
</blockquote></details>

Changed



_Code quote:_
```PHP
$routingKey = strtoupper($bitOasisCurrency);
```

**Review by foowie — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bit-oasis/bitfinex-api/12)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @RobMKR)
